Document migrating end devices from v2

neoaggelos opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes #1684
Closes #1685 

#### Changes
<!-- What are the changes made in this pull request? -->

- Write migration guide.
- This is the documentation part. It depends on https://github.com/TheThingsNetwork/ttn/pull/776, a new ttnctl binary release, and a new v3 release with https://github.com/TheThingsNetwork/lorawan-stack/pull/2457

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

@benolayinka There are probably quite a few inconsistencies in writing style, probably some missing information
@htdvisser @johanstokking Scope and correctness, especially in the Major Changes page.
@laurensslats Scope, probably coordinate testing as well.

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md`. The target branch is set to `master` if the changes are fully compatible with existing API, database, configuration and CLI.
- [X] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [X] Documentation: Relevant documentation is added or updated.
- [X] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

benolayinka commented:
Fixed styling issues and added ttnv2 shortcode. Also moved existing tts shortcode to the-things-stack theme

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttn-lw-cli.md:
```diff
@@ -0,0 +1,31 @@
+description: Install and configure CLI tools
+---
+
+You will also need the latest version of `ttn-lw-cli`, the CLI for {{% tts %}}.
+
+Get the latest version of `ttn-lw-cli` by following instructions for the [Command-line Interface]({{< ref "getting-started/cli" >}}).
+
+
```
Use existing CLI shortcode

johanstokking replied:
Also remove double empty lines

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttn-lw-cli.md:
````diff
@@ -0,0 +1,31 @@
+Configure `ttn-lw-cli` for connecting to {{% tts %}}.
+
+```bash
+# for Cloud Hosted:
+$ ttn-lw-cli use "<tenant-id>.<region>.cloud.thethings.industries"
````
This does not belong in TTS

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttn-lw-cli.md:
```diff
@@ -0,0 +1,31 @@
+# for Cloud Hosted:
+$ ttn-lw-cli use "<tenant-id>.<region>.cloud.thethings.industries"
+
+# or for a private deployment on `thethings.example.com`
+# (if you followed the the Getting Started guide):
+$ ttn-lw-cli use "thethings.example.com"
```
Assume that people followed the getting started guide in TTS

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttn-lw-cli.md:
````diff
@@ -0,0 +1,31 @@
+$ ttn-lw-cli login --callback=false
+```
+
+> **NOTE**: The commands above assume that ttn-lw-cli can be found in your PATH. If that
+> is not the case, you need to specify the path to `ttn-lw-cli` instead, e.g. `./ttn-lw-cli`.
````
This should all be covered by the CLI shortcode and the referred instructions for using the CLI. Let's avoid the sliding edge where we mention these notes on each page where we use the CLI.

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttn-lw-cli.md:
````diff
@@ -0,0 +1,31 @@
+Login to the {{% tts %}} with the next command, and follow the instructions.
+
+```bash
+$ ttn-lw-cli login --callback=false
````
Why no callback?

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttn-lw-cli.md:
```diff
@@ -0,0 +1,31 @@
+---
+title: Configure ttn-lw-cli
```
Maybe here it's better to call it `Configure V3 CLI`

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttnctl.md:
```diff
@@ -0,0 +1,32 @@
+---
+title: Configure ttnctl
```
And here `Configure V2 CLI`

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttnctl.md:
```diff
@@ -0,0 +1,32 @@
+[3]: https://accounts.thethingsnetwork.com
+[4]: https://account.thethingsnetwork.org/users/authorize?client_id=ttnctl&redirect_uri=/oauth/callback/ttnctl&response_type=code
+
+> **NOTE**: The commands above assume that ttnctl can be found in your PATH. If that
+> is not the case, you need to specify the path to `ttnctl` instead, e.g. `./ttnctl`.
```
Remove

johanstokking commented on doc/content/getting-started/migrating-from-v2/create-application.md:
```diff
@@ -0,0 +1,18 @@
+
+## Create a New Application on {{% tts %}}
+
+Create a new application on {{% tts %}} where the end devices will be
+imported. This can be done from by following instructions for [Creating an Application in the Console]({{% ref "getting-started/console/create-application" %}})
+or [Creating an Application using the CLI]({{% ref "getting-started/cli#create-application" %}}).
```
Remove line breaks

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttnctl.md:
```diff
@@ -0,0 +1,32 @@
+description: Install and configure CLI tools
+---
+
+You will need to use the latest version of `ttnctl`, the CLI for {{% ttnv2 %}}.
+Follow the [instructions from the TTN documentation][1]. An overview is given below:
```
Remove line breaks

johanstokking replied:
This is not addressed

johanstokking commented on doc/content/getting-started/migrating-from-v2/create-application.md:
```diff
@@ -0,0 +1,18 @@
+> **NOTE**: In {{% tts %}}, applications do not have an `AppEUI`. The `AppEUI` is
+> configured per-device.
+
+> **NOTE**: If creating the application using the CLI, do not forget to
+> [link with the Network Server]({{% ref "getting-started/cli#link-application" %}}).
```
This should be covered by the aforementioned docs on creating an application with CLI, right?

johanstokking commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
```diff
@@ -0,0 +1,50 @@
+---
+title: Export End Devices from V2
```
Inconsistent casing

johanstokking commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
```diff
@@ -0,0 +1,50 @@
+
+## Export End Devices with ttnctl
+
+We will export the end devices of this application in a JSON format that can then
+be parsed and imported by {{% tts %}}.
```
I don't think we write guides in the "we" form, are we?

johanstokking commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
```diff
@@ -0,0 +1,50 @@
+We will export the end devices of this application in a JSON format that can then
+be parsed and imported by {{% tts %}}.
+
+For the commands below, we are using an example application on The Things Network
+with **AppID** `v2-application` and **AppEUI** `0011223300112233`.
```
Line breaks, please check further occurrences

johanstokking commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
````diff
@@ -0,0 +1,50 @@
+$ ttnctl devices export "DeviceID" --frequency-plan-id EU_863_870 > device.json
+```
+
+Alternatively, you can export all the end devices with a single command and save
+them in `all-devices.json`.
+
+```bash
+$ ttnctl devices export "DeviceID" --frequency-plan-id EU_863_870 > all-devices.json
````
This seems like one device still, though

johanstokking commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
```diff
@@ -0,0 +1,50 @@
+> means that all end devices need a frequency plan. Above, we specify
+> the `EU_863_870` frequency plan, but this needs to be replaced with the
+> frequency plan corresponding to your region. A list of supported frequency plan IDs in
+> [the lorawan-frequency-plans Github repository](https://github.com/TheThingsNetwork/lorawan-frequency-plans/blob/master/frequency-plans.yml).
```
I think we should make sure this is in the docs too

johanstokking replied:
Also mind the line breaks here

johanstokking commented on doc/content/getting-started/migrating-from-v2/import-devices.md:
````diff
@@ -0,0 +1,26 @@
+in the previous step with `ttn-lw-cli`:
+
+```bash
+$ ttn-lw-cli dev create --application-id "v3-application" < devices.json
````
Wasn't it `imported-application` ?

johanstokking commented on doc/content/getting-started/migrating-from-v2/import-devices.md:
```diff
@@ -0,0 +1,26 @@
+This will import your devices on {{% tts %}}. In case
+any device fails, you see a relevant error message at the end of the output.
+
+> **ΝΟΤΕ**: After importing an end device to {{% tts %}}, you should remove it
+> from {{% ttnv2 %}}. For OTAA devices, it is enough to simply change the
+> AppKey, so the device can no longer connect but the existing session is preserved.
+> Next time the device joins, it will connect to {{% tts %}}.
+>
+> Keep in mind that an end device can only be registered in one Network Server
+> at a time.
```
Provide instructions for this. This is crucial

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,94 @@
+{{% ttnv2 %}} and {{% tts %}}, along with some guidelines to make the
+migration process easier to manage.
+
+### Application Data
```
I don't think we should start with this. I would go from architecture, to gateway, network to application-layer changes.

neoaggelos replied:
Updated, not sure how much we want to write about the changes that are not entirely related to migrating existing applications and workloads.

johanstokking replied:
This is heading `##` and MQTT, payload format, storage etc is al `###` (like now)

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttnctl.md:
```diff
@@ -0,0 +1,29 @@
+
+[1]: https://www.thethingsnetwork.org/docs/network/cli/quick-start.html
+[2]: https://www.thethingsnetwork.org/docs/network/cli/quick-start.html#installation
+[3]: https://accounts.thethingsnetwork.com
```
This URL doesn't seem right

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttnctl.md:
````diff
@@ -0,0 +1,29 @@
+$ ttnctl selfupdate
+```
+
+Go to [https://accounts.thethingsnetwork.com][3] and click [ttnctl access code][4].
````
Can we skip [3] here?

johanstokking commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
````diff
@@ -0,0 +1,40 @@
+$ ttnctl devices export-all --frequency-plan-id EU_863_870 > all-devices.json
+```
+
+> **NOTE**: You can find a list of supported frequency plan IDs in [the lorawan-frequency-plans Github repository](https://github.com/TheThingsNetwork/lorawan-frequency-plans/blob/master/frequency-plans.yml).
````
This should be a reference

See #2421

benolayinka replied:
I thought by 'no hugo magic' that you meant let's just continue to link to github. we DO want to manually copy the frequency plans to the reference section?

neoaggelos replied:
Leaving the link to the GitHub repo for now, with a TODO comment to link to our documentation page after #2421 has been merged.

johanstokking replied:
OK I'm fine with a TODO comment as long as there is a (specific) open issue that's referenced

johanstokking commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
```diff
@@ -0,0 +1,40 @@
+
+> **NOTE**: The exported end devices contain the device name, description, location data, activation mode (ABP/OTAA), root keys and the application AppEUI.
+
+> **NOTE**: For OTAA end devices, the session keys are not preserved. After importing, the end device will need to send a new join request on the Network Server of {{% tts %}}.
```
Wait, can't we transfer the session keys?

neoaggelos replied:
Slipped on that one. The comment should have been about MAC state, https://github.com/TheThingsNetwork/ttn/pull/776#discussion_r418002025.

johanstokking commented on doc/content/getting-started/migrating-from-v2/import-devices.md:
```diff
@@ -0,0 +1,27 @@
+weight: 50
+---
+
+## Import Devices using ttn-lw-cli
```
Remove header; add CLI only shortcode

johanstokking commented on doc/content/getting-started/migrating-from-v2/import-devices.md:
```diff
@@ -0,0 +1,27 @@
+
+## Import Devices using ttn-lw-cli
+
+To import your devices, use the `devices.json` file you created
+in the previous step with `ttn-lw-cli`:
```
Remove line breaks, check other occurrences

johanstokking replied:
This is not addressed

johanstokking commented on doc/content/getting-started/migrating-from-v2/import-devices.md:
````diff
@@ -0,0 +1,27 @@
+
+```bash
+# set to 0000000000000000, or use any other 64-bit hex string
+$ ttnctl devices set "DeviceID" --app-key "0000000000000000"
````
There's no "disable" counterpart like `export-all`, is there?

Do we want that?

We certainly don't want conflicting join-accepts to be sent

neoaggelos replied:
We do not have that currently. It might make sense to have that, but then we would not want users to lose their AppKey's by mistake.

johanstokking replied:
Yes, that is to be avoided anyway.

How about rewriting the instructions like:

1. Export one or all devices to a file (as backup + input for 3)
2. Reset AppKey of one or all devices
3. Pipe the file to the V3 CLI end devices create command

neoaggelos replied:
Moved the section for changing the AppKey after exporting.

johanstokking replied:
OK, but still it doesn't work for multiple devices at once.

It is not feasible to reset the AppKey one-by-one.

neoaggelos replied:
Should this be added on ttnctl, or can we just tell the users to do something like:

```bash
devices=$(ttnctl devices list | grep -v "INFO" | grep -v "\<DevID.*AppEUI.*DevEUI.*DevAddr.*Description\>" | cut -f1)

for dev in $devices; do
  ttnctl devices update ....
done
```

neoaggelos replied:
An alternative would be to add a `ttnctl devices disable` (and a `disable-all` version) that resets AppKey's of devices and keeps the old one as a device attribute (e.g. `ttnctl-backup-appkey`).

Possibly this is getting too far, but that would allow us to have a `re-enable` command, that restores the AppKey from that attribute.

Are attributes safe to use for that, or are they public? If they are, we could also print the old AppKey and tell the user to store it somewhere safe.

johanstokking replied:
Promoting the discussion, see below

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,77 @@
+
+Before getting started, we will discuss major breaking changes between {{% ttnv2 %}} and {{% tts %}}, along with some guidelines to make the migration process easier to manage.
+
+### Architecture
```
I think we need `##` here

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,77 @@
+weight: 10
+---
+
+Before getting started, we will discuss major breaking changes between {{% ttnv2 %}} and {{% tts %}}, along with some guidelines to make the migration process easier to manage.
```
Add `more` break after this

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,77 @@
+
+### LoRaWAN support
+
+{{% tts %}} requires the LoraWAN MAC Protocol Version and Regional Parameters(LoraWAN PHY version) to be set per device. These default to `MAC_V1_0_2` and
```
Spacing and line breaks

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,77 @@
+{{% tts %}} requires the LoraWAN MAC Protocol Version and Regional Parameters(LoraWAN PHY version) to be set per device. These default to `MAC_V1_0_2` and
+`MAC_PHY_V1_0_2_REV_B` for devices imported from {{% ttnv2 %}}.
+
+The LoraWAN MAC settings are configurable per end device. This means that all end devices need a frequency plan. Above, we used the `EU_863_870` frequency plan, but this needs to be replaced with the frequency plan corresponding to your region. A list of supported frequency plan IDs in [the lorawan-frequency-plans Github repository](https://github.com/TheThingsNetwork/lorawan-frequency-plans/blob ...
```
Need reference

neoaggelos replied:
Reference to what?

johanstokking replied:
TODO for #2421

neoaggelos replied:
Added

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,77 @@
+
+The LoraWAN MAC settings are configurable per end device. This means that all end devices need a frequency plan. Above, we used the `EU_863_870` frequency plan, but this needs to be replaced with the frequency plan corresponding to your region. A list of supported frequency plan IDs in [the lorawan-frequency-plans Github repository](https://github.com/TheThingsNetwork/lorawan-frequency-plans/blob ...
+
+Furthermore, {{% tts %}} brings full support for all LoraWAN versions (1.0.0 through 1.1.0), as well as Class B and Class C devices.
```
Casing

I would remove the version range as it's subject to change and sometimes versions get inserted (like 1.0.4 while there was 1.0.3 and 1.1 already)

neoaggelos replied:
Added list of versions instead. Maybe we can drop this too, and just say "All LoRaWAN versions"?

johanstokking replied:
I prefer the latter indeed

neoaggelos replied:
Updated accordingly

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,77 @@
+
+Furthermore, {{% tts %}} brings full support for all LoraWAN versions (1.0.0 through 1.1.0), as well as Class B and Class C devices.
+
+### Gateway coverage
```
Title Case

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,77 @@
+
+The Packet Broker enables peering between networks, so traffic received by one network (e.g. the public community network) but intended for a different network ({{% tts %}}) can be forwarded to and from that network.
+
+With Packet Broker enabled (which is not discussed here), you can receive traffic on {{% tts %}} without the need to configure any of your gateways (since  traffic will be automatically forwarded from the public community network).
```
You might want to refer to the Peering guide / concept

neoaggelos replied:
I am not sure which one that is.

johanstokking replied:
There's one under Integrations now

neoaggelos replied:
It is under the `develop` branch and this guide is based on master, so it is not available yet. Link will work after these are merged, is this enough? Should I rebase to develop?

johanstokking replied:
Yeah retarget to `develop`; we're releasing this anyway

johanstokking replied:
This is now up-to-date

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,77 @@
+
+{{% tts %}} also supports HTTP webhooks and PubSubs for nats.io or MQTT.
+
+### Payload formats
```
Title Case

neoaggelos commented:
Updated according to comments, but I have asked for some clarification on a few things. Also added a section for people that want to configure the MAC settings of their devices, I thinik it will be super useful for people new to the v3.

benolayinka commented:
Sylvain and Rish put this in confluence - I don't know if it's public or how much is still accurate, but some of the information could be helpful in the docs. 

Q. How does migration from V2 to V3 SaaS work?
A. We move applications and devices from V2 to V3 at a scheduled time; devices cannot be present in both. This means that the user has to adapt their connector to both V2 and V3 to avoid downtime. Device sessions will be preserved, however as V3 has more MAC settings, manual configuration may be required in the NS.

Gateways need to be updated to a new address, preferably migrating to Basic Station protocol. We may provide a migration time for gateways and route traffic from the V2 bridge to the V3 environment so that they can serve both for a while (although this is without guarantees as you can get conflicting downlink scheduling and duty cycle violations).


Q. What kind of MAC settings do I need to adjust, and how does it affect my data?
A. In most cases, you don’t need to adjust anything. V3 has per-device settings for LoRaWAN MAC and PHY version (i.e. setting the device to LoRaWAN 1.0.2 with Regional Parameters 1.0.2 rev B). On migration, we assume V2 defaults when migrating to V3. Devices that work in V2 will work in V3 as well. However, you may want to adjust RX1 timing and RX2 parameters for better performance.

Q. Which device session settings are you migrating?
A. Device ID, AppEUI, DevEUI, DevAddr, NwkSKey, AppSKey, FCntUp, FCntDown and PHY as bare minimum.

Q. Do I migrate application per application?
A. Yes. You cannot migrate per device, only applications as a whole.

Q. Can I migrate devices back to V2?
A. No.

Q. How to migrate my gateway to V3?
A. We will offer a temporary bridging solution so that gateways work on both your existing V2 SaaS environment and the new V3 SaaS environment. After that, we will update DNS so that they will point to the new environment exclusively. However, it is highly recommended to upgrade to the Gateway Agent with V3 gateway ID and API key, or update the address when using UDP.

Q. How will I know if my gateway supports the Basic Station?
A. The basic station is maintained by Semtech so most gateway will support in the future. The Things Indoor Gateway and The Things Outdoor gateway support it, for example. MultiTech and Kerlink will also support it. For more information about a specific model, contact the manufacturer.

Q. Can I modify the frequency plan or channel plan (ex: Rx windows) ?
A. Yes. You can make your own channel plan (channel frequencies, rx windows, ...) plan within a frequency plan (EU823, AU915,...). If you are hosting V3 by yourself you will be able to define your own frequency plan. This feature will come to cloud hosted (TTI SaaS) deployment later.

Q. Can I separate my own customer into different tenant (multi tenancy) ?
A. In our cloud offering each of our customer is a tenant of the TTI network. To add one of your own customer you can simply request another tenant from TTI.
Self Hosted and Dedicated Host deployments will have access to the tenant management API.

Q. Will the APIs change in V3?
A. Yes. The APIs will change, please find the V3 API Documentation here. 

Q. Do you see any possibility to leverage the platform stack for NB-IoT?
A. NB-IoT: No
    LTE-M: Possibly

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -17,13 +17,18 @@ Behind the scenes, {{% tts %}} is using a new architecture. See {{< ref "/refere
+The Packet Broker enables peering between networks, so traffic received by one network (e.g. the public community network) but intended for a different network ({{% tts %}}) can be forwarded to and from that network. See the [Peering Guide]({{< ref "/integrations/peering" >}}) for details on Packet Broker and how to enable it for your network.
 
-With Packet Broker enabled (which is not discussed here), you can receive traffic on {{% tts %}} without the need to configure any of your gateways (since  traffic will be automatically forwarded from the public community network).
+With Packet Broker enabled, you can receive traffic on {{% tts %}} without the need to configure any of your gateways (since  traffic will be automatically forwarded from the public community network).
```
Remove last sentence; that is only in The Things Industries Cloud Hosted

neoaggelos commented:
It is not too straightforward to just retarget to `develop` because it brings a few changes from master as well. Rebasing does not help either, because the commits from master are preserved. I think the easiest way would be to create a new branch based on develop and cherry pick these commits, then force push to the current branch. But is it worth it?

neoaggelos commented:
> It is not too straightforward to just retarget to `develop` because it brings a few changes from master as well. Rebasing does not help either, because the commits from master are preserved. I think the easiest way would be to create a new branch based on develop and cherry pick these commits, then force push to the current branch. But is it worth it?

Went ahead and did it anyway

benolayinka commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,83 @@
+
+## Architecture
+
+Behind the scenes, {{% tts %}} is using a new architecture. See {{< ref "/reference/components" >}} for a complete overview of the different components of {{% tts %}}.
```
See `[Components]({{< ref "/reference/components" >}})` for a complete overview

neoaggelos replied:
Fixed

johanstokking commented:
@neoaggelos 

What I think we should address is that we do have tooling to export all devices, but we do not have tooling to reset the AppKey (or SKeys for ABP) all devices. This means that most users will need to reset the AppKey one-by-one, or don't bother with that, and hope they're lucky with a join-accept from the right LNS.

So, a few solutions:

1. A bash script. This is a no go as it's not user friendly and not portable
2. A command to reset the AppKey (or SKeys) of all end devices. This is something we can do
3. Add a flag to `export` and `export-all` to reset the concerning keys. Also, change the doc here to write to a file, so that data doesn't get lost while piping and a failing V3 CLI command

neoaggelos commented:
> 2. A command to reset the AppKey (or SKeys) of all end devices. This is something we can do

I think this is the cleanest solution.

htdvisser commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,83 @@
+
+{{% tts %}} has support for an uplink payload formatter (similar to the payload decoder) and a downlink payload formatter (similar to the payload encoder). These can be set per application, and can even be overridden per end device.
+
+Migrating the {{% ttnv2 %}} payload encoder and decoder to an uplink and downlink payload formatter should be straightforward, since they have the same format.
```
I think we need to link to a reference here, because we now also have some environment in addition to the payload and f_port arguments.

neoaggelos replied:
Added a TODO comment with a reference to https://github.com/TheThingsNetwork/lorawan-stack/issues/598, since payload formatters are not yet documented.

neoaggelos commented on doc/themes/the-things-stack/layouts/shortcodes/ttnv2.html:
```diff
@@ -0,0 +1 @@
+The Things Network Stack V2
```
cc: @benolayinka for this one.

benolayinka replied:
I moved it because it seems like a basic element of TTS that doesn't depend on anything site specific, and it could migrate cleanly to another site as part of the theme.

neoaggelos commented on doc/content/getting-started/migrating-from-v2/configure-mac-settings.md:
````diff
@@ -0,0 +1,30 @@
+MAC Parameters on {{% tts %}} are configurable per end device. After you have imported your end devices, you can use the CLI to configure their MAC settings.
+
+```bash
+$ ttn-lw-cli devices update --help
````
Updated

neoaggelos commented on doc/content/getting-started/migrating-from-v2/configure-ttn-lw-cli.md:
```diff
@@ -0,0 +1,22 @@
+
+{{< cli-only >}}
+
+Configure `ttn-lw-cli` for connecting to {{% tts %}}.
```
The `{{< cli-only >}}` shortcode links to the CLI installation guide. The parts below were just a quick way to setup the CLI with the v3 where the devices will be migrated. The point was to not break the flow of the guide too much. Removed since it's not required.

neoaggelos commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
````diff
@@ -0,0 +1,50 @@
+
+```bash
+# set to 0000000000000000, or use any other 64-bit hex string
+$ ttnctl devices set "DeviceID" --app-key "0000000000000000"
````
Updated this part to use the `convert-to-abp` command.

neoaggelos commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
````diff
@@ -0,0 +1,83 @@
+
+{{% tts %}} uses a different data format for uplink and downlink traffic than {{% ttnv2 %}}.
+
+Example uplink message on {{% ttnv2 %}}:
+
+```js
+/* topic name: app1/devices/dev1/up */
+{"app_id":"app1","dev_id":"dev1","hardware_serial":"1122334411223344","port":1,"counter":0,"payload_raw":"EQ==","payload_fields":{"led":17},"metadata":{"time":"2020-05-01T00:04:41.258830149Z","latitude":47.984,"longitude":43.123,"altitude":100}}
+```
+
+Example downlink message on {{% ttnv2 %}}:
+
+```js
+/* topic name: app1/devices/dev1/down */
+{"port":10,"confirmed":true,"payload_raw":"EBA="}
+```
````
I decided not to include a test v3 message here, as it is already well documented on the linked page for data formats.

johanstokking replied:
I wonder if it's relevant to show the V2 format here. That is also extensively documented and you can point to that

neoaggelos replied:
Removed and added link to ttn docs

neoaggelos commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,83 @@
+
+### Payload Formats
+
+{{% ttnv2 %}} has support for payload decoders, converters, validators (for uplink) and encoders (for downlink). These can be either CayenneLP or Javascript functions.
```
fixed

johanstokking commented:
@neoaggelos why is this targeting `develop`?

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-mac-settings.md:
```diff
@@ -0,0 +1,30 @@
+
+{{< cli-only >}}
+
+This section is optional.
```
Optional for what?

neoaggelos replied:
removed

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-mac-settings.md:
```diff
@@ -0,0 +1,30 @@
+
+This section is optional.
+
+MAC Parameters on {{% tts %}} are configurable per end device. After you have imported your end devices, you can use the CLI to configure their MAC settings.
```
Isn't this generic for all devices, not just migrated devices?

rvolosatovs replied:
For ABP this has to be done during import, otherwise MAC settings are not effective

neoaggelos replied:
@rvolosatovs Then we have to set this when exporting the devices from the v2? Is this documented somewhere? It would have been helpful to know this in advance.

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-mac-settings.md:
````diff
@@ -0,0 +1,30 @@
+$ ttn-lw-cli end-devices update "AppID" "DeviceID" --unset mac-settings.rx1-delay
+```
+
+This way you will be able to fine tune your devices so that they work as expected with the Network Server of {{% tts %}}.
````
I understand the goal of this guide. I would suggest changing it from something generic about MAC state, to a more optional but recommended guide about finetuning MAC state for V3.

That would definitely include changing the RX1 delay to 5 seconds.

@rvolosatovs is there any other desired MAC parameter that we want to mention here?

rvolosatovs replied:
Nothing specific to migration, I think we should just have a generic list of v3 MAC settings with use cases and reference it here.

neoaggelos replied:
Will create an issue and add a TODO comment for this.

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttnctl.md:
```diff
@@ -0,0 +1,27 @@
+description: Install and configure CLI tools
+---
+
+You will need to use the latest version of `ttnctl`, the CLI for {{% ttnv2 %}}. Follow the [instructions from the TTN documentation][1]. An overview is given below:
```
Expand TTN

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttnctl.md:
````diff
@@ -0,0 +1,27 @@
+$ ttnctl selfupdate
+```
+
+Go to https://account.thethingsnetwork.org and click [ttnctl access code][3].
````
This URL gets rendered as text.

Please locally run the docs and look at the pages before asking review

neoaggelos replied:
Had removed link after it was suggested previously. Markdown rendered as link so I assumed Hugo did the same. Re-added link

johanstokking commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
````diff
@@ -0,0 +1,57 @@
+In order to export a single device, use the following command. The device will be saved to `device.json`.
+
+```bash
+$ ttnctl devices export "DeviceID" --frequency-plan-id EU_863_870 > device.json
````
We don't support capitals in device IDs, so `DeviceID` is a weird example

neoaggelos replied:
changed to `device_id` (and `app_id`)

johanstokking commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
```diff
@@ -0,0 +1,57 @@
+Add reference to docs after that is merged.
+-->
+
+> **NOTE**: The exported end devices contain the device name, description, location data, activation mode (ABP/OTAA), root keys and the application AppEUI. It also contains the session keys, so your OTAA devices can simply keep working with {{% tts %}}.
```
This should be in the intro of this guide

neoaggelos replied:
Moved at the top.

johanstokking commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
```diff
@@ -0,0 +1,57 @@
+
+> **NOTE**: The exported end devices contain the device name, description, location data, activation mode (ABP/OTAA), root keys and the application AppEUI. It also contains the session keys, so your OTAA devices can simply keep working with {{% tts %}}.
+
+> **IMPORTANT**: Keep in mind that an end device can only be registered in one Network Server at a time. After importing an end device to {{% tts %}}, you should remove it from {{% ttnv2 %}}. For OTAA devices, it is enough to simply change the AppKey, so the device can no longer connect but the existing session is preserved. Next time the device joins, it will connect to {{% tts %}}.
```
Afaik we don't use the `**IMPORTANT**` block anywhere.

Isn't there a box in the theme for special attention, i.e. warning level or something?

neoaggelos replied:
I don't think we have something like that.

neoaggelos replied:
Changed to `NOTE` instead of `IMPORTANT`

johanstokking commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
```diff
@@ -0,0 +1,57 @@
+
+> **IMPORTANT**: Keep in mind that an end device can only be registered in one Network Server at a time. After importing an end device to {{% tts %}}, you should remove it from {{% ttnv2 %}}. For OTAA devices, it is enough to simply change the AppKey, so the device can no longer connect but the existing session is preserved. Next time the device joins, it will connect to {{% tts %}}.
+
+After exporting, make sure to clear the AppKey of your OTAA devices. This can be achieved easily with the following command:
```
Remove `easily`

johanstokking commented on doc/content/getting-started/migrating-from-v2/import-devices.md:
````diff
@@ -0,0 +1,18 @@
+$ ttn-lw-cli end-devices create --application-id "imported-application" < devices.json
+```
+
+This will import your devices on {{% tts %}}. In case
+any device fails, you see a relevant error message at the end of the output.
````
Same and other occurrences

neoaggelos replied:
hopefully no newlines are left now

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,90 @@
+
+## Architecture
+
+Behind the scenes, {{% tts %}} is using a new architecture. See [Components]({{< ref "/reference/components" >}}) for a complete overview of the different components of {{% tts %}}.
```
Remove "behind the scenes", this is not a show

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,90 @@
+
+## LoRaWAN support
+
+{{% tts %}} requires the LoRaWAN MAC Protocol Version and Regional Parameters (LoRaWAN PHY version) to be set per end device. These default to `LoRaWAN 1.0.2` and `LoRaWAN Regional Parameters 1.0.2 Rev B` for end devices imported from {{% ttnv2 %}}, because this configuration is the most consistent with the V2.
```
"LoRaWAN version"

johanstokking replied:
The mentioned versions are not code, so don't put literal styling

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,90 @@
+
+The Packet Broker enables peering between networks, so traffic received by one network (e.g. the public community network) but intended for a different network ({{% tts %}}) can be forwarded to and from that network. See the [Peering Guide]({{< ref "/integrations/peering" >}}) for details on Packet Broker and how to enable it for your network.
+
+With Packet Broker enabled, you can receive traffic on {{% tts %}} without the need to configure any of your gateways.
```
This is only if the old network is also on Packet Broker. That is not the case for most V2 private networks. So I would claim this only when the gateways are connected to The Things Network public community network

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,90 @@
+
+[Read more about using the MQTT Server]({{< ref "/integrations/mqtt" >}}).
+
+{{% tts %}} also supports HTTP webhooks and PubSubs for nats.io or MQTT.
```
- Point to integrations tab
- NATS
- MQTT server

Why is this mentioned under the "MQTT Traffic" section

neoaggelos replied:
Added separate Integrations section with reference link to the docs page.

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,90 @@
+
+{{% tts %}} also supports HTTP webhooks and PubSubs for nats.io or MQTT.
+
+### Payload Formats
```
This should be above MQTT

johanstokking replied:
If you care about alphabetic order, make working with application data a larger section, start with format, and then talk about what has changed per integration

neoaggelos replied:
Moved

neoaggelos commented:
> @neoaggelos why is this targeting `develop`?

https://github.com/TheThingsNetwork/lorawan-stack/pull/2467#discussion_r420656529
https://github.com/TheThingsNetwork/lorawan-stack/pull/2467#issuecomment-624553161

neoaggelos commented:
Updated many parts of the migration guide. Will squash commits after the review process is done.

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,82 @@
+
+With Packet Broker enabled on both {{% tts %}} and {{% ttnv2 %}}, you can receive traffic on {{% tts %}} without having to re-configure any of your gateways.
+
+> **NOTE**: Packet Broker is enabled on The Things Network Public Community Network.
```
+ and The Things Industries Cloud Hosted

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,82 @@
+Add a reference here once Payload Formatters are documented.
+-->
+
+### MQTT Traffic
```
MQTT is no different than webhooks; please add a section to that as well

The payload format changed as well as downlink endpoints, paths and security settings. Also, it's now called Webhooks (it was HTTP integration)

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,82 @@
+
+{{% tts %}} does not currently support a Storage integration similar to {{% ttnv2 %}}. This feature will be added in a future release.
+
+## Suggested migration process
```
Title case

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,82 @@
+
+First, update applications to support the {{% tts %}} data format. If you are using payload formatters, make sure to set them correctly from the Application settings page.
+
+Follow the rest of the guide and start by migrating a small number of test end devices (and gateways, if needed) to {{% tts %}}. Once you are confident that your devices are working properly, migrate the rest of your devices and gateways.
```
Seeing a first, not seeing a second

Probably we want to advice people to setup a device on V3 first and test everything first, before migrating existing devices in the field

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-mac-settings.md:
```diff
@@ -0,0 +1,43 @@
+
+### Setting RX1 Delay
+
+The RX1 delay of end devices is set to 1 second by default. For some end devices, this may lead to traffic being dropped. In that case, it is recommended that the RX1 delay be increased to 5 seconds:
```
For some end devices, this may lead to downlink messages not being scheduled in time. Therefore, ...

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-mac-settings.md:
````diff
@@ -0,0 +1,43 @@
+$ ttn-lw-cli end-devices update "app-id" "device-id" --mac-settings.rx1-delay RX_DELAY_5
+```
+
+> *NOTE*: For ABP end devices, you may have to reset the device for the changes to take effect.
````
Again, what does this mean?

If the ABP device resets this state, and the RX1 delay mismatches with the NS MAC state, we don't have a downlink path at all. I would argue that this setting is not applicable to ABP devices at all, as the RX1 delay is part of the personalization. For ABP, you have to set this setting both in NS and end device.

rvolosatovs replied:
> If the ABP device resets this state, and the RX1 delay mismatches with the NS MAC state, we don't have a downlink path at all

Network Server will reset stored device MAC state and get the correct RX1 delay so we will, in fact, have a downlink path.
But indeed, in majority of cases `mac-settings.rx1-delay` MUST be correct at *creation* time of the device and updating this is quite tricky(possible though, by changing `mac-state.current-parameters.rx1-delay` *and* `mac-settings.rx1-delay`)

johanstokking commented on doc/content/getting-started/migrating-from-v2/configure-ttn-lw-cli.md:
```diff
@@ -0,0 +1,11 @@
+
+{{< cli-only >}}
+
+> *NOTE*: Make sure that you have followed the linked installation guide and have successfully logged in to {{% tts %}}.
```
`**NOTE**`, fix other occurrences is needed

johanstokking commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
```diff
@@ -0,0 +1,59 @@
+
+> **NOTE**: Keep in mind that an end device can only be registered in one Network Server at a time. After importing an end device to {{% tts %}}, you should remove it from {{% ttnv2 %}}. For OTAA devices, it is enough to simply change the AppKey, so the device can no longer join but the existing session is preserved. Next time the device joins, the activation will be handled by {{% tts %}}.
+
+### Disable Exported End Devices on the V2
```
-the

johanstokking commented on doc/content/getting-started/migrating-from-v2/export-v2-devices.md:
````diff
@@ -0,0 +1,59 @@
+$ ttnctl devices convert-all-to-abp --save-to-attribute "original-app-key"
+```
+
+> **NOTE** The AppKey of each device will be printed on the standard output, and stored as a device attribute (with name `original-app-key`). You can retrieve the device attributes with `ttnctl devices info "device-id"`.
````
`**NOTE**:`

johanstokking commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,90 @@
+
+## LoRaWAN support
+
+{{% tts %}} requires the LoRaWAN Version and Regional Parameters (LoRaWAN PHY version) to be set per end device. These default to LoRaWAN version 1.0.2 and LoRaWAN Regional Parameters version 1.0.2 Rev B for end devices imported from {{% ttnv2 %}}, because this configuration is the most consistent with the V2.
```
version

benolayinka commented on doc/content/getting-started/migrating-from-v2/major-changes.md:
```diff
@@ -0,0 +1,90 @@
+
+{{% tts %}} does not currently support a Storage integration similar to {{% ttnv2 %}}. This feature will be added in a future release.
+
+## Suggested Migration Process
```
Maybe break these steps up using headers:

<img width="886" alt="Screenshot 2020-05-11 at 10 58 07" src="https://user-images.githubusercontent.com/6963436/81543579-8a10c400-9376-11ea-8a19-7604dd668e68.png">

```
**First:** Update applications to support the {{% tts %}} data format. If you are using payload formatters, make sure to set them correctly from the Application settings page.

**Then:** Follow this guide in order to migrate a single end device (and gateway, if needed) to {{% tts %}}. Continue by gradually migrating your end devices in small batches. Avoid migrating production workloads before you are certain that they will work as expected.

**Finally:** Once you are confident that your end devices are working properly, migrate the rest of your devices and gateways to {{% tts %}}.
```

laurensslats commented:
I assumed this was part of the v3.8.2 release, but apparently not.

As customers are chasing us on a daily base, we cannot wait for another 2 weeks before adding this documentation to our website.

Any thoughts on how/when we can publish this document?

@KrishnaIyer @neoaggelos
